runtime/buffer.h

```diff
@@ -334,51 +334,41 @@ inline bool can_fuse(const dim& inner, const dim& outer) {
   return inner.stride() * inner.extent() == outer.stride();
 }
 
-template <typename F>
-void for_each_contiguous_slice(void* base, const dim* dims, int d, index_t elem_size, const F& f,
-    index_t slice_extent = 1, index_t extent = 1) {
-  if (d == -1) {
-    // We've handled all the loops, call the function.
-    f(base, slice_extent);
-    return;
-  }
-
-  const slinky::dim& dim = dims[d];
-  extent *= dim.extent();
-  if (extent <= 0) {
-    // Don't want to worry about empty dimensions in the cases below.
-    return;
-  }
+struct for_each_contiguous_slice_dim {
+  union {
+    // For loop_folded to call flat_offset_bytes
+    const slinky::dim* dim;
+    // For loop_linear to offset the base.
+    index_t stride;
+  };
+  index_t begin;
+  index_t extent;
+  enum {
+    call_f,
+    loop_linear,
+    loop_folded,
+  } impl;
+};
 
-  index_t stride = dim.stride();
-  if (stride == elem_size) {
-    // This is the dense dimension, pass this dimension through.
-    assert(slice_extent == 1);  // Two dimensions of stride == elem_size...?
-    assert(dim.min() / dim.fold_factor() == dim.max() / dim.fold_factor());
-    if (d > 0) {
-      for_each_contiguous_slice(base, dims, d - 1, elem_size, f, extent);
-    } else {
-      f(base, extent);
-    }
-  } else if (d > 0 && can_fuse(dims[d - 1], dim)) {
-    // This dimension can be fused with the next dimension.
-    for_each_contiguous_slice(base, dims, d - 1, elem_size, f, slice_extent, extent);
-  } else if (dim.fold_factor() == dim::unfolded) {
-    for (index_t i = 0; i < extent; ++i, base = offset_bytes(base, stride)) {
-      for_each_contiguous_slice(base, dims, d - 1, elem_size, f, slice_extent);
+template <typename F>
+void for_each_contiguous_slice(void* base, const for_each_contiguous_slice_dim* dim, const F& f) {
```

Comment:
style nit: we use `dim` as the name for a `class dim` everywhere else; using it here for a different type leads to things like `dim->dim` which (to my eyes) is confusing. Not sure of a better name but we already have overloaded `dim` to mean `a class for dimensions` and/or `an instance of class dim` so adding a third overloaded meaning is maybe not ideal